src/coreComponents/mesh/generators/VTKMeshGenerator.cpp

```diff
@@ -323,270 +304,210 @@ std::vector< T > gatherData( std::vector< T > const & data )
-  fillCellBlockNames( regionsTetra, ElementType::Tetrahedron );
-  fillCellBlockNames( regionsWedges, ElementType::Prism );
-  fillCellBlockNames( regionsPyramids, ElementType::Pyramid );
-
-  // Communicate all the region names
-  // TODO if in a mesh, there is for instance a region with tetra, and a region without tetra,
-  //      they will both contain a cell block tetrahedron
-  std::vector< int > const allCellBlockRegionIndex = gatherData( cellBlockRegionIndex );
-  std::vector< ElementType > const allCellBlockElementType = gatherData( cellBlockElementType );
-  for( std::size_t i = 0; i < allCellBlockRegionIndex.size(); i++ )
-  {
-    for( std::size_t j = 0; j < allCellBlockElementType.size(); j++ )
-    {
-      switch( allCellBlockElementType[j] )
+      switch( dim )
       {
-        case ElementType::Tetrahedron:
-          regionsTetra[allCellBlockRegionIndex[i]];
-          break;
-        case ElementType::Hexahedron:
-          regionsHex[allCellBlockRegionIndex[i]];
+        case 0:
+        case 1:
+        {
+          // Ignore vertex/line elements for now; maybe later we can import well polylines here
           break;
-        case ElementType::Prism:
-          regionsWedges[allCellBlockRegionIndex[i]];
+        }
+        case 2:
+        {
+          // Merge all 2D elements together as polygons (we don't track their shapes).
+          std::vector< vtkIdType > & surfaceCells = elemsByType[ ElementType::Polygon ];
+          surfaceCells.insert( surfaceCells.end(), cellsByTypeTemp[t].begin(), cellsByTypeTemp[t].end() );
           break;
-        case ElementType::Pyramid:
-          regionsPyramids[allCellBlockRegionIndex[i]];
+        }
+        case 3:
+        {
+          // Collect 3D elements as is
+          elemsByType.emplace( type, std::move( cellsByTypeTemp[t] ) );
           break;
+        }
         default:
-          GEOSX_ERROR( allCellBlockElementType[j] << " is not a recognized cell type" );
+          break;
       }
     }
   }
+  return elemsByType;
+}
 
-  // We want to know the surface attributes from all MPI ranks.
-  // Then we'll be able to allocate an entry in the `surfacesIdsToCellsIds` table.
-  // The `cellsIds` can possibly be empty (if the rank does not hold any cells for the surface).
-  std::vector< int > surfaces;
-  surfaces.reserve( surfacesIdsToCellsIds.size() );
-  for( auto const & s2c: surfacesIdsToCellsIds )
+VTKMeshGenerator::CellMapType
+splitElementsByTypeAndAttribute( std::unordered_map< ElementType, std::vector< vtkIdType > > & elemsByType,
+                                 vtkDataArray * const attributeDataArray )
+{
+  VTKMeshGenerator::CellMapType elemsByTypeAndAttribute;
+  for( auto & t2c : elemsByType )
   {
-    surfaces.push_back( s2c.first );
-  }
+    ElementType const elemType = t2c.first;
+    std::vector< vtkIdType > & cells = t2c.second;
+    std::unordered_map< int, std::vector< vtkIdType > > & elemsByAttribute = elemsByTypeAndAttribute[elemType];
```

Comment:
I could also do an additional pass to count cells for each value of `attribute`. There will still be some allocations in that pass (an `unordered_map` is required to store counts, since the set or range of `attribute` values are not known in advance), but probably a lot fewer than when filling vectors from 0 capacity.

Reply:
@jeannepellerin do you have any advice about this?